File: assignments/springboot/jpa-board/src/main/java/com/example/jpaboard/service/BoardService.java
package com.example.jpaboard.service;

import com.example.jpaboard.domain.entity.Board;
import com.example.jpaboard.domain.repository.BoardRepository;
import com.example.jpaboard.dto.BoardDeleteRequestDto;
import com.example.jpaboard.dto.BoardUpdateRequestDto;
import com.example.jpaboard.exception.BoardNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.time.LocalDateTime;
import java.util.List;
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class BoardService {
    private final BoardRepository boardRepository;
    private final FileService fileService;

    public List<Board> getBoardList(int page, int size) {
        Pageable pageable = PageRequest.of(page-1, size, Sort.by("id").descending());

        return boardRepository.findAll(pageable).getContent();
    }

    public int getTotalBoards() {
        return (int) boardRepository.count();
    }

    public Board getBoardDetail(Long id) {
        return boardRepository.findById(id)
                .orElseThrow(() -> new BoardNotFoundException("게시글을 찾을 수 없음 id=" + id));
    }

    @Transactional
    public void saveArticle(String userId, String title, String content, MultipartFile file) {
        String filePath = fileService.storeFile(file);

        Board board = Board.builder()
                .userId(userId)
                .title(title)
                .content(content)
                .filePath(filePath)
                .created(LocalDateTime.now())
                .build();

        boardRepository.save(board);
    }

    @Transactional
    public void updateArticle(Long id, BoardUpdateRequestDto request) {
        Board board = boardRepository.findById(id)
                .orElseThrow(() -> new BoardNotFoundException("게시글을 찾을 수 없음 id=" + id));
        String filePath = board.getFilePath();

        if (request.isFileFlag()) {
            fileService.deleteFile(board.getFilePath());
            filePath = fileService.storeFile(request.getFile());
        }

        board.update(request.getTitle(), request.getContent(), filePath);
    }

    @Transactional
    public void deleteArticle(Long id, BoardDeleteRequestDto request) {
        if (!boardRepository.existsById(id)) {
            throw new BoardNotFoundException("게시글을 찾을 수 없음 id=" + id);
        }

        boardRepository.deleteById(id);
        fileService.deleteFile(request.getFilePath());
    }
}

File: practice/springboot/board/kotlin-board/src/main/kotlin/com/example/kotlinboard/service/BoardService.kt
package com.example.kotlinboard.service

import com.example.kotlinboard.domain.entity.Board
import com.example.kotlinboard.domain.repository.BoardRepository
import com.example.kotlinboard.dto.BoardCreateRequest
import com.example.kotlinboard.dto.BoardUpdateRequest
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class BoardService(private val repository: BoardRepository) {
    fun getBoards(page: Int, size: Int, keyword: String?): Page<Board> {
        val pageable: Pageable = PageRequest.of(page - 1, size, Sort.by("id").descending())

        return if (keyword.isNullOrBlank()) {
            repository.findAll(pageable)
        } else {
            repository.findByTitleContaining(keyword, pageable)
        }
    }

    // 게시글 작성
    @Transactional
    fun createBoard(request: BoardCreateRequest) {
        repository.save(request.toEntity())
    }

    // 게시글 상세 조회
    fun getBoard(id: Long): Board? = repository.findByIdOrNull(id)

    // 트랜잭션 안에서 조회한 엔티티는 영속 상태라,
    // 값을 바꾸면 트랜잭션이 끝날 때 JPA가 변경을 감지해 UPDATE SQL을 자동으로 날림 (dirty checking)
    @Transactional
    fun updateBoard(id: Long, request: BoardUpdateRequest) {
        // ?: -> 엘비스 연산자 [null이면 수행]
        val board = repository.findByIdOrNull(id) ?: return

        board.update(request.title, request.content)
    }

    @Transactional
    fun deleteBoard(id: Long) {
        repository.deleteById(id)
    }
}